brightray/browser/browser_context.cc

```diff
@@ -88,20 +70,23 @@ BrowserContext::BrowserContext(const std::string& partition, bool in_memory)
 
   content::BrowserContext::Initialize(this, path_);
 
+  io_handle_ = new URLRequestContextGetter::Handle(GetWeakPtr());
+
   browser_context_map_[PartitionKey(partition, in_memory)] = GetWeakPtr();
 }
 
 BrowserContext::~BrowserContext() {
   DCHECK_CURRENTLY_ON(BrowserThread::UI);
   NotifyWillBeDestroyed(this);
   ShutdownStoragePartitions();
-  if (BrowserThread::IsMessageLoopValid(BrowserThread::IO)) {
-    BrowserThread::DeleteSoon(BrowserThread::IO, FROM_HERE,
-                              resource_context_.release());
-    BrowserThread::PostTask(
-        BrowserThread::IO, FROM_HERE,
-        base::BindOnce(&URLRequestContextGetter::NotifyContextShutdownOnIO,
-                       base::RetainedRef(url_request_getter_)));
+  io_handle_->ShutdownOnUIThread();
+}
+
+void BrowserContext::OnDestruct() const {
+  if (BrowserThread::CurrentlyOn(BrowserThread::UI)) {
+    delete this;
```

Comment:
Usually `RefcountedThreadSafe` classes use the default destruction trait https://cs.chromium.org/chromium/src/base/memory/ref_counted.h?l=313, but we have specified a custom delete to make sure `BrowserContext` gets destroyed on the UI thread, since we have references on the IO thread and can't guarantee where the final release can happen.